File: answers/other/in sap/Path Sum.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def hasPathSum(self, root, sum):
        """
        :type root: TreeNode
        :type sum: int
        :rtype: bool
        """
      	
        return (None != root and None == root.left and None == root.right and root.val == sum) or (None != root and (self.hasPathSum(root.left, sum - root.val) or self.hasPathSum(root.right, sum - root.val)))

File: answers/leetcode/Path Sum/Path Sum.cpp
//@result 114 / 114 test cases passed. Status: Accepted Runtime: 12 ms Submitted: 0 minutes ago You are here! Your runtime beats 12.82% of cpp submissions.

/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool hasPathSum(TreeNode *root, int sum) {
        // Start typing your C/C++ solution below
        // DO NOT write int main() function
        if (NULL == root)
        {
			return false;
		}
		if (sum == root->val && NULL == root->left && NULL == root->right)
		{
			return true;
		}
		if (hasPathSum(root->left, sum - root->val) || hasPathSum(root->right, sum - root->val))
		{
			return true;
		}
		return false;
    }
};
